File: src/main/java/com/wowo/wowo/domain/shared/AggregateRoot.java
package com.wowo.wowo.domain.shared;

import java.util.List;

/**
 * Base interface for aggregate roots in the domain.
 * Aggregates are clusters of domain objects that can be treated as a single unit.
 */
public interface AggregateRoot {
    
    /**
     * Gets the list of domain events that have been raised by this aggregate.
     */
    List<DomainEvent> getDomainEvents();
    
    /**
     * Marks all domain events as handled by clearing the event list.
     */
    void markEventsAsHandled();
    
    /**
     * Adds a domain event to be published when the aggregate is saved.
     */
    void addDomainEvent(DomainEvent event);
}

File: src/main/kotlin/com/wowo/wowo/shared/domain/AggregateRoot.kt
package com.wowo.wowo.shared.domain

/**
 * Base class for Aggregate Roots in DDD
 */
abstract class AggregateRoot<ID> : Entity<ID>() {
    private val domainEvents: MutableList<DomainEvent> = mutableListOf()

    fun addDomainEvent(event: DomainEvent) {
        domainEvents.add(event)
    }

    fun clearDomainEvents() {
        domainEvents.clear()
    }

    fun getDomainEvents(): List<DomainEvent> = domainEvents.toList()
}


